### PR TITLE
CAMEL-8010:Locking the critical section to avoid race condition if Ag…

### PR DESCRIPTION
**Pull request by rajithapl**
…gregateTimeOutChecker also completes at the same time as Recover task

### CODE REVIEW

File: camel-core/src/main/java/org/apache/camel/processor/aggregate/AggregateProcessor.java [hunk 1]
```diff
@@ -1223,68 +1223,77 @@ public void run() {
...
-                            // send to DLC
-                            try {
                                 // set redelivery counter
                                 exchange.getIn().setHeader(Exchange.REDELIVERY_COUNTER, data.redeliveryCounter);
-                                exchange.getIn().setHeader(Exchange.REDELIVERY_EXHAUSTED, Boolean.TRUE);
-                                deadLetterProducerTemplate.send(recoverable.getDeadLetterUri(), exchange);
-                            } catch (Throwable e) {
-                                exchange.setException(e);
-                            }
+                                if (recoverable.getMaximumRedeliveries() > 0) {
```
Comment (onders86) [inline camel-core/src/main/java/org/apache/camel/processor/aggregate/AggregateProcessor.java]: i am not sure if you can give up setting redelivery_exhausted setting in exhange.header. that would be feature change. @davsclaus @oscerd @zregvart could better comment..

Comment (rajithapl) [inline camel-core/src/main/java/org/apache/camel/processor/aggregate/AggregateProcessor.java]: I haven't changed those part of the code. I added only the locking part. line # 1226 -1229 and line# 1292 - 1296. I didnt touch any of the other lines. Must be just formatting changes due to the additional try block

 if(!optimisticLocking){
                    lock.lock();
                }
               try {
-----------------------------------
}
                finally {
                   if(!optimisticLocking){
                       lock.unlock();
                 }
               }

Comment (onders86) [inline camel-core/src/main/java/org/apache/camel/processor/aggregate/AggregateProcessor.java]: yes, probable formatting confusion on my side. LG.
I would suggest running full core unit tests. if there is not an issue, i guess this PR may go into further releases but i guess still this is not a full resolution to CAMEL-8010 issue?

Comment (rajithapl) [inline camel-core/src/main/java/org/apache/camel/processor/aggregate/AggregateProcessor.java]: Thanks. Unit test look fine. Yes. This is to handle race condition between AggregateTimeoutChecker and AggregateRecoverChecker calling onSubmitCompletion at the same time. I found one other potential placement where we don't use lock. AggregateProcessor.doProcess() line #354 - 358 should go inside the locked block. I tried that but it failed the Junit AggregatorLockingTest. Would it be better if I put this fix against a separate jira?

**Comment by davsclaus** [comment]
Can you fix checkstyle with

    mvn compile -P sourcecheck

**Comment by rajithapl** [comment]
Done

**Comment by oscerd** [comment]
Thanks the PR has been merged. Can you close this?

**Comment by rajithapl** [comment]
Thanks

**Pull request closed without merging**